### ISSUE TITLE
Tracking issue for builtin declarative macros

### ISSUE THREAD

**Issue by edwin0cheng** [rust-analyzer/rust-analyzer#2212]
Here is the full list of declarative macros current rustc [supported](https://github.com/rust-lang/rust/blob/fa0f7d0080d8e7e9eb20aa9cbf8013f96c81287f/src/libsyntax_ext/lib.rs#L64):

* [ ] `asm`: [asm::expand_asm](https://github.com/rust-lang/rust/blob/e8f43b72eb2596f360004f6cdf9cdde4b9e789e1/src/libsyntax_ext/asm.rs#L43)
* [ ] `assert`: [assert::expand_assert](https://github.com/rust-lang/rust/blob/a3b6e5705cff9c69362b7ed2d273ffc148b564db/src/libsyntax_ext/assert.rs#L13)
* [ ] `cfg`: [cfg::expand_cfg](https://github.com/rust-lang/rust/blob/e8f43b72eb2596f360004f6cdf9cdde4b9e789e1/src/libsyntax_ext/cfg.rs#L14)
* [ ] `column`: [source_util::expand_column](https://github.com/rust-lang/rust/blob/a3b6e5705cff9c69362b7ed2d273ffc148b564db/src/libsyntax_ext/source_util.rs#L33)
* [ ] `compile_error`: [compile_error::expand_compile_error](https://github.com/rust-lang/rust/blob/fa0f7d0080d8e7e9eb20aa9cbf8013f96c81287f/src/libsyntax_ext/compile_error.rs#L7)
* [ ] `concat_idents`: [concat_idents::expand_concat_idents](https://github.com/rust-lang/rust/blob/e8f43b72eb2596f360004f6cdf9cdde4b9e789e1/src/libsyntax_ext/concat_idents.rs#L11)
* [ ] `concat`: [concat::expand_concat](https://github.com/rust-lang/rust/blob/50f8aadd746ebc929a752e5ffb133936ee75c52f/src/libsyntax_ext/concat.rs#L8)
* [ ] `env`: [env::expand_env](https://github.com/rust-lang/rust/blob/e8b190ac4ad79e58d21ee1d573529ff74d8eedaa/src/libsyntax_ext/env.rs#L46)
* [ ] `file`: [source_util::expand_file](https://github.com/rust-lang/rust/blob/a3b6e5705cff9c69362b7ed2d273ffc148b564db/src/libsyntax_ext/source_util.rs#L46)
* [ ] `format_args_nl`: [format::expand_format_args_nl](https://github.com/rust-lang/rust/blob/5a5027519a4a634baa6cde5b698b907d27fbe6b3/src/libsyntax_ext/format.rs#L856)
* [ ] `format_args`: [format::expand_format_args](https://github.com/rust-lang/rust/blob/5a5027519a4a634baa6cde5b698b907d27fbe6b3/src/libsyntax_ext/format.rs#L848)
* [ ] `global_asm`: [global_asm::expand_global_asm](https://github.com/rust-lang/rust/blob/e8f43b72eb2596f360004f6cdf9cdde4b9e789e1/src/libsyntax_ext/global_asm.rs#L22)
* [ ] `include_bytes`: [source_util::expand_include_bytes](https://github.com/rust-lang/rust/blob/a3b6e5705cff9c69362b7ed2d273ffc148b564db/src/libsyntax_ext/source_util.rs#L157)
* [ ] `include_str`: [source_util::expand_include_str](https://github.com/rust-lang/rust/blob/a3b6e5705cff9c69362b7ed2d273ffc148b564db/src/libsyntax_ext/source_util.rs#L126)
* [ ] `include`: [source_util::expand_include](https://github.com/rust-lang/rust/blob/a3b6e5705cff9c69362b7ed2d273ffc148b564db/src/libsyntax_ext/source_util.rs#L157)
* [x] `line`: [source_util::expand_line](https://github.com/rust-lang/rust/blob/a3b6e5705cff9c69362b7ed2d273ffc148b564db/src/libsyntax_ext/source_util.rs#L22)
* [ ] `log_syntax`: [log_syntax::expand_log_syntax](https://github.com/rust-lang/rust/blob/fa0f7d0080d8e7e9eb20aa9cbf8013f96c81287f/src/libsyntax_ext/log_syntax.rs#L6)
* [ ] `module_path`: [source_util::expand_mod](https://github.com/rust-lang/rust/blob/a3b6e5705cff9c69362b7ed2d273ffc148b564db/src/libsyntax_ext/source_util.rs#L61)
* [ ] `option_env`: [env::expand_option_env](https://github.com/rust-lang/rust/blob/e8b190ac4ad79e58d21ee1d573529ff74d8eedaa/src/libsyntax_ext/env.rs#L14)
* [ ] `stringify`: [source_util::expand_stringify](https://github.com/rust-lang/rust/blob/a3b6e5705cff9c69362b7ed2d273ffc148b564db/src/libsyntax_ext/source_util.rs#L55)
* [ ] `trace_macros`: [trace_macros::expand_trace_macros](https://github.com/rust-lang/rust/blob/fa0f7d0080d8e7e9eb20aa9cbf8013f96c81287f/src/libsyntax_ext/trace_macros.rs#L6)

After #2205 is landed, some of the builtin declarative macros are: 

Trivial to implement:
* `stringify`, `column`, `compile_error`, etc

Trivial to implement for IDE purpose: (A dummy implement will work here)
* `asm`, `global_asm`, etc

Need  `OUTDIR`  metadata support:
* `env`, etc

Eager macro expansion support:
`concat`, `include`, etc

**Comment by edwin0cheng** [comment]
After #2205 is landed, some of the builtin declarative macros are: 

Trivial to implement:
* `stringify`, `column`, `compile_error`, etc

Trivial to implement for IDE purpose: (A dummy implementation will work here)
* `asm`, `global_asm`, etc

Need  `OUTDIR`  metadata support:
* `env`, etc

Need eager macro expansion support:
`concat`, `include`, etc

**Comment by matklad** [comment]
Priority wise, handling `include!(concat!(env!("OUT_DIR"), "/foo.rs"))` would be sweet!

**Comment by edwin0cheng** [comment]
I will take `concat!` to see how to support eager expansion in current infrastructure. 

But I still can't figure it out how name resolution works inside inner expansion phase 🤣

**Comment by kjeremy** [comment]
@edwin0cheng what does `compile_error!` expand to? Just a string with the error contents?

**Comment by edwin0cheng** [comment]
Um... something as `!` result type?  e.g `loop { "Compiler Error" }` for now ? @matklad

**Comment by matklad** [comment]
Yeah, I think `loop {}` is good

**Comment by Tehnix** [comment]
NOTE: This might be out of scope for this issue, so feel free to let me know and I'll create a separate issue.
---
With Rust 1.54 [recently introducing support](https://blog.guillaume-gomez.fr/articles/2021-08-03+Improvements+for+%23%5Bdoc%5D+attributes+in+Rust#Using%20doc%20comments%20to%20extend%20tests%20capabilities) for `#[doc = include_str!("compile_fail.rs")]` I was curious if the "support" that is aim for here also includes expanded these macros in the generated docs on hover?

Take for example the following code:

```rust
/// The GraphQL schema for our mock service.
///
/// ```graphql
#[doc = include_str!("schema.graphql")]
/// ```
pub struct Query;
```

It pulls in an external file, `schema.graphql`, which includes the full schema of the `Query`. At the moment, the tooltip in VS Code will ignore the `#[doc = include_str!("schema.graphql")]` part in the documentation it shows:

<img width="411" alt="Screen Shot 2021-08-03 at 21 49 35" src="https://user-images.githubusercontent.com/1189998/128084336-ac9c90bf-1eae-4eb2-a8de-e433d4d0e2bc.png">

If you run `cargo doc --open` it does show up in the final documentation though:
<img width="819" alt="Screen Shot 2021-08-03 at 21 53 19" src="https://user-images.githubusercontent.com/1189998/128084743-0dd5cced-b426-4744-b5b5-64661d267d14.png">
